File: src-cpp/leetcode/best-time-to-buy-and-sell-stock.cpp
#include <bits/stdc++.h>
using namespace std;

int maxProfit(vector<int> &prices)
{
    int profit = 0;
    int minPrice = INT_MAX;

    for (auto price : prices)
    {
        minPrice = min(minPrice, price);
        profit = max(profit, price - minPrice);
    }

    return profit;
}

// Alter
int maxProfit(vector<int> &prices)
{
    int maxProfitSoFar = 0;
    int currProfit = 0;

    for (int i = 1; i < prices.size(); i += 1)
    {
        currProfit = max(0, currProfit + (prices[i] - prices[i - 1]));
        maxProfitSoFar = max(maxProfitSoFar, currProfit);
    }

    return maxProfitSoFar;
}

File: src-ts/leetcode/best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
  let minPrice = INT_MAX;
  let profit = 0;

  for (const price of prices) {
    minPrice = Math.min(minPrice, price);
    profit = Math.max(profit, price - minPrice);
  }

  return profit;
}

// Alternative

function maxProfitAlter(prices: number[]): number {
  let maxProfitSoFar = 0;
  let currProfit = 0;

  for (let i = 1; i < prices.length; i += 1) {
    currProfit = Math.max(0, currProfit + (prices[i] - prices[i - 1]));

    maxProfitSoFar = Math.max(maxProfitSoFar, currProfit);
  }

  return maxProfitSoFar;
}
